[SVG] animateMotion accumulate doesn't work properly with rotate: auto / auto-reverse

Ahmad-S792 opened the pull request:
#### b3b752433b78753c0744f4fed28036380c512c6b
<pre>
[SVG] animateMotion accumulate doesn&apos;t work properly with rotate: auto / auto-reverse

<a href="https://bugs.webkit.org/show_bug.cgi?id=256928">https://bugs.webkit.org/show_bug.cgi?id=256928</a>
rdar://problem/109489241

Reviewed by Simon Fraser.

This patch aligns WebKit behavior with Gecko / Firefox and Blink / Chromium.

Merge: <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=154375

We need not accumulate the angle for animation as we have a method to get normal angle at a given length on the path.
So, only position needs to be accumulated. Also made some changes to simplify the logic for accumulation.

* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(SVGAnimateMotionElement::buildTransformForProgress): Remove &apos;angle&apos; accumulation from here
(SVGAnimateMotionElement::calculateAnimatedValue): Move here
* LayoutTests/svg/animations/animateMotion-accumulate-1a.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-1a-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-1b.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-1b-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-1c.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-1c-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-2a.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-2a-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-2b.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-2b-expected.svg: Add Test Case Expectation

All tests have slight pixel tolerances to accommodate for platform differences.

Canonical link: <a href="https://commits.webkit.org/264595@main">https://commits.webkit.org/264595@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9460ac38a6b59c1318f13d00119f54e720f504e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8056 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8341 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8554 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/9708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/8157 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8060 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10327 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8252 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/9708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8200 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/9282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7344 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/9832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/6583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/7372 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/9832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7706 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/7496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/9832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7967 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/7280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/11489 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/966 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/7708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/svg/SVGAnimateMotionElement.cpp:
```diff
@@ -240,6 +233,16 @@ void SVGAnimateMotionElement::calculateAnimatedValue(float percentage, unsigned
+    float positionOnPath = m_animationPath.length() * percentage;
+    auto traversalState(m_animationPath.traversalStateAtLength(positionOnPath));
+
+    float angle = traversalState.normalAngle();
```
Not obvious to the reader that `angle` is in degrees, but that's not new.

Ahmad-S792 replied:
Should I add comment or rename (e.g., 'degreeAngle') and then just do 'unsafe-merge' or it is to leave as is?

Ahmad-S792 replied:
`// The 'angle' below is in 'degrees'.` <- Added this comment. @smfr

Ahmad-S792 committed:
[SVG] animateMotion accumulate doesn't work properly with rotate: auto / auto-reverse

https://bugs.webkit.org/show_bug.cgi?id=256928
rdar://problem/109489241

Reviewed by Simon Fraser.

This patch aligns WebKit behavior with Gecko / Firefox and Blink / Chromium.

Merge: https://src.chromium.org/viewvc/blink?view=revision&revision=154375

We need not accumulate the angle for animation as we have a method to get normal angle at a given length on the path.
So, only position needs to be accumulated. Also made some changes to simplify the logic for accumulation.

* Source/WebCore/svg/SVGAnimateMotionElement.cpp:
(SVGAnimateMotionElement::buildTransformForProgress): Remove 'angle' accumulation from here
(SVGAnimateMotionElement::calculateAnimatedValue): Move here
* LayoutTests/svg/animations/animateMotion-accumulate-1a.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-1a-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-1b.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-1b-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-1c.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-1c-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-2a.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-2a-expected.svg: Add Test Case Expectation
* LayoutTests/svg/animations/animateMotion-accumulate-2b.svg: Add Test Case
* LayoutTests/svg/animations/animateMotion-accumulate-2b-expected.svg: Add Test Case Expectation

All tests have slight pixel tolerances to accommodate for platform differences.

Canonical link: https://commits.webkit.org/264595@main
